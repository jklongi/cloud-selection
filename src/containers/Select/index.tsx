import { useState } from 'react'
import MuiSelect, { SelectProps } from '@material-ui/core/Select'
import MenuItem from '@material-ui/core/MenuItem'
import Input from '@material-ui/core/Input'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'

type Item<V> = {
    name: string
    value: V
}

type Props<V> = Omit<SelectProps, 'onChange'> & {
    id: string
    label: string
    items: Item<V>[]
    onChange: (selected: V[]) => void
}

const Select = <V extends string>({ items, onChange, id, label, ...rest }: Props<V>) => {
    const [value, setValue] = useState<V[]>([])
    const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
        const value = event.target.value as V[]
        setValue(value)
        onChange(value)
    }
    return (
        <FormControl fullWidth>
            <InputLabel htmlFor={id}>{label}</InputLabel>
            <MuiSelect
                multiple
                value={value}
                onChange={handleChange}
                fullWidth
                input={<Input />}
                inputProps={{ id }}
                {...rest}
            >
                {items.map(({ name, value }) => (
                    <MenuItem key={value} value={value}>
                        {name}
                    </MenuItem>
                ))}
            </MuiSelect>
        </FormControl>
    )
}

export default Select
