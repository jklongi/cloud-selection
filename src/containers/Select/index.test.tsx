import { mount } from 'enzyme'
import MenuItem from '@material-ui/core/MenuItem'
import Select from './index'

const items = [
    { name: 'First', value: '1' },
    { name: 'Second', value: '2' },
    { name: 'Third', value: '3' },
]

describe('<Select />', () => {
    it('matches snapshot', () => {
        const wrapper = mount(<Select items={items} onChange={() => {}} label="Label" id="select" />)
        expect(wrapper).toMatchSnapshot()
    })
    it('renders label', () => {
        const wrapper = mount(<Select items={items} onChange={() => {}} label="Label" id="select" />)
        expect(wrapper.find('label').text()).toEqual('Label')
    })
    it('changes fire callback', () => {
        const onChange = jest.fn()
        const wrapper = mount(<Select items={items} onChange={onChange} label="Label" id="select" />)
        wrapper.find('[role="button"]').simulate('mousedown', { button: 0 })
        wrapper.find(MenuItem).at(2).simulate('click')
        expect(onChange).toHaveBeenCalled()
    })

    it('changes provides correct callback value with single select', () => {
        let state: string[] = []
        const onChange = (selected: string[]) => {
            state = selected
        }
        const wrapper = mount(<Select items={items} onChange={onChange} label="Label" id="select" />)
        wrapper.find('[role="button"]').simulate('mousedown', { button: 0 })
        wrapper.find(MenuItem).at(2).simulate('click')
        expect(state).toEqual(['3'])
    })
    it('changes provides correct callback value with multiple select', () => {
        let state: string[] = []
        const onChange = (selected: string[]) => {
            state = selected
        }
        const wrapper = mount(<Select items={items} onChange={onChange} label="Label" id="select" />)
        wrapper.find('[role="button"]').simulate('mousedown', { button: 0 })
        console.log(wrapper.find(MenuItem).at(2).simulate('click'))
        console.log(wrapper.find(MenuItem).at(0).simulate('click'))
        expect(state).toEqual(['3', '1'])
    })
})
