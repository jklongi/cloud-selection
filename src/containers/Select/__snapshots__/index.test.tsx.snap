// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> matches snapshot 1`] = `ReactWrapper {}`;
